File: master/divide_int_without_operator.py

class Solution:
    # @return an integer
    def divide(self, dividend, divisor):
        sign = (dividend<0 and divisor<0) or (dividend>0 and divisor>0)
        res = 0
        dividend = dividend*(1 if dividend>=0 else -1)
        divisor = divisor*(1 if divisor>=0 else -1)
        if dividend<divisor:
            return 0
        for i in range(31,-1,-1):
            if(dividend>>i >= divisor):
                res += 1<<i
                dividend -= (divisor<<i)
        return (1 if sign else -1)*res

print Solution().divide(1,0)

File: 5/divide_int_without_operator.cpp
/*

Divide two integers without using multiplication, division and mod operator.

If it is overflow, return MAX_INT.

*/

#include "header.h"

int divide(int dividend, int divisor) {
    if(dividend == 0) return 0;
    bool flag = dividend>=0 && divisor>=0 || dividend<=0 && divisor<=0;
    if(divisor == 0) return flag?INT_MAX:INT_MIN;
    unsigned int dd = dividend>=0?dividend:-1*dividend;
    unsigned int ds = divisor>=0?divisor:-1*divisor;
    if(dd<ds) return 0;
    unsigned int res = 0;
    for(int i=31; i>=0; i--){
        if((dd>>i) >= ds){
            res += (unsigned int)(1<<i);
            dd -= (ds<<i);
        }
    }
    if(res > INT_MAX) return flag?INT_MAX:INT_MIN;
    else return flag?res:-1*res;
}

int main(){
//    printf("%d\n", divide(-999511578, -2147483648));
//    printf("%d\n", divide(100579234, -555806774));
    cout<<sizeof(long)<<" "<<sizeof(int)<<endl;
    printf("%d\n", divide(-2147483648, -1));
    printf("%d\n", divide(-2147483648, 1));
//    printf("%d\n", divide(5, 2));
//    printf("%d\n", divide(1, -1));
    return 0;
}
